File: leetcode/222-count-complete-tree-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
 public:
  int countNodes(TreeNode* root)
  {
    int sum = 0;
    for (TreeNode* node = root; node; ) {
      int lh = 0, rh = 0;
      for (TreeNode* n = node->left; n; n = n->left) ++lh;
      for (TreeNode* n = node->right; n; n = n->left) ++rh;
      if (lh > rh) {
        sum += 1 << rh;
        node = node->left;
      } else {
        sum += 1 << lh;
        node = node->right;
      }
    }
    return sum;
  }
};

File: leetcode/222-count-complete-tree-nodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def _height(node):
            h = 0
            n = node
            while n is not None:
                h += 1
                n = n.left
            return h

        cnt = 0
        node = root
        while node is not None:
            lh = _height(node.left)
            rh = _height(node.right)
            if lh > rh:
                cnt += 1 << rh
                node = node.left
            else:
                cnt += 1 << lh
                node = node.right
        return cnt
